File: app/src/main/java/com/example/wagtailjournal/Container.java
package com.example.wagtailjournal;

import java.util.HashMap;

public class Container {
    private HashMap<String, ServiceFactory> factories = new HashMap();
    private HashMap<String, Object> services = new HashMap();

    public Container defserv(String name, ServiceFactory sf) {
        factories.put(name, sf);
        return this;
    }

    public Object geti(String name) {
        if (services.containsKey(name)) {
            return services.get(name);
        }
        ServiceFactory sf = factories.get(name);
        services.put(name, sf.create(this));
        return services.get(name);
    }
}

File: app/src/main/java/com/github/tkojitu/wagtailjournal/Container.kt
package com.github.tkojitu.wagtailjournal

class Container {
    private val factories = HashMap<String, ServiceFactory>()
    private val services = HashMap<String, Any?>()

    fun defserv(name: String, sf: ServiceFactory): Container {
        factories[name] = sf
        return this
    }

    fun geti(name: String): Any? {
        if (services.containsKey(name)) {
            return services[name]
        }
        val sf = factories[name]
        services[name] = sf!!.create(this)
        return services[name]
    }
}
